### ISSUE TITLE
lightdm can't start a GNOME wayland session

### ISSUE THREAD

**Issue by worldofpeace** [NixOS/nixpkgs#56342]
## Issue description
Mostly opening for investigation, someone in IRC mentioned that they couldn't
get a wayland session to start when using lightdm, as they noticed this when `$XDG_SESSION_TYPE == x11`.

I was then able to reproduce this within a VM.

### Steps to reproduce
1. Enable GNOME3 and lightdm as a display-manager
2. Try to start the `GNOME` session
3. Realize that you're running the gnome-xorg session

---
cc @hedning @jtojnar

**Comment by hedning** [comment]
Yeah, I didn't turn it on in the gnome wayland pr as lightdm didn't handle the duplicate `gnome.desktop` sesssion names:
https://github.com/NixOS/nixpkgs/blob/f3a676a88cdde63862a8cf7c232fc4f6e7eddaef/nixos/modules/services/x11/display-managers/lightdm.nix#L48

https://github.com/NixOS/nixpkgs/pull/53843 turns it on (note this pr needs to be updated for pantheon).

**Comment by worldofpeace** [comment]
> Yeah, I didn't turn it on in the gnome wayland pr as lightdm didn't handle the duplicate gnome.desktop sesssion names

Thanks for the heads up

> #53843 turns it on (note this pr needs to be updated for pantheon).

I'll push some fixup commits to that branch so it can work for pantheon

**Comment by hedning** [comment]
Cheers :)

According to upstream it's up to the specific greeter to handle duplicates: https://github.com/CanonicalLtd/lightdm/issues/16

**Comment by lePerdu** [comment]
This isn't specific to LightDM: GDM does the same thing for me. I don't know how all of the XSession stuff works, but I looked at the .desktop files in `/nix/store/...gnome-session...` (`xsessions/gnome.desktop`, `xsessions/gnome-xorg.desktop`, and `wayland-sessions/gnome.desktop`) and found them identical except for the "Name" field. Am I missing something, or should the "Exec" fields (and possibly others) be different for the XOrg and Wayland sessions?

**Comment by hedning** [comment]
Yeah, there's some behind the scenes stuff being done depending on where the desktop files reside (ie. `wayland-sesisons/` vs. `xsessions/`)

> GDM still runs the X session for both in NixOS

Works here. `nixos-18.09` doesn't have gnome wayland support, that might be what you're seeing. Otherwise    I would double check that `services.xserver.displayManager.gdm.wayland` is `true` (should be by default). There might be some hardware thing too (eg. nvidia card with the proprietary drivers, don't think we support that yet).

**Comment by lePerdu** [comment]
I'm on the `nixos-unstable` channel and can start the GNOME wayland session manually with
```
env XDG_SESSION_TYPE=wayland dbus-run-session /nix/store/...gnome-session-3.28.1/bin/gnome-session
```

**Comment by worldofpeace** [comment]
LightDM can now start wayland sessions, it needed a patch upstream that I provided
* https://github.com/NixOS/nixpkgs/commit/2cb421afb977e2be7ca8694996c1fb39200ead03

However it is broken with GNOME still. Haven't had time to debug it.

**Comment by hedning** [comment]
I'm guessing lightdm finds the xsession first (since there's a gnome.desktop in both). Should probably work to just add the wayland session dir before the xsession one. Checking now.

**Comment by worldofpeace** [comment]
Oh, I forgot to mention how when I tested, I worked around the duplicate session name bug (which plasma5 actually compensates for in sddm, gnome could do similar).
I removed the xsessions from gnome-session in testing to be sure.

**Comment by hedning** [comment]
Ahh, right, using the wayland dir first worked, in the sense that it tries to launch a wayland session. But something is indeed broken as it won't start the session.

**Comment by worldofpeace** [comment]
still an issue

**Comment by tasmo** [comment]
Still not working.

**Comment by MohamedLEGH** [comment]
still not working.

**Comment by qweered** [comment]
Gnome is wayland only now, if anyone still have this issue please comment
